sdk-libs/compressed-token-sdk/src/decompress_runtime.rs

```diff
@@ -0,0 +1,160 @@
+    type Accounts<'info>;
+
+    /// Get seeds for the token account PDA (used for decompression).
+    fn get_seeds<'a, 'info>(
+        &self,
+        accounts: &'a Self::Accounts<'info>,
+        remaining_accounts: &'a [AccountInfo<'info>],
+    ) -> Result<(Vec<Vec<u8>>, Pubkey), ProgramError>;
+
+    /// Get authority seeds for signing during compression.
+    fn get_authority_seeds<'a, 'info>(
+        &self,
+        accounts: &'a Self::Accounts<'info>,
+        remaining_accounts: &'a [AccountInfo<'info>],
+    ) -> Result<(Vec<Vec<u8>>, Pubkey), ProgramError>;
+}
+
+/// Token decompression processor.
+#[inline(never)]
+#[allow(clippy::too_many_arguments)]
+pub fn process_decompress_tokens_runtime<'info, 'a, 'b, V, A>(
+    accounts_for_seeds: &A,
+    remaining_accounts: &[AccountInfo<'info>],
+    fee_payer: &AccountInfo<'info>,
+    ctoken_program: &AccountInfo<'info>,
+    ctoken_rent_sponsor: &AccountInfo<'info>,
+    ctoken_cpi_authority: &AccountInfo<'info>,
+    ctoken_config: &AccountInfo<'info>,
+    config: &AccountInfo<'info>,
+    ctoken_accounts: Vec<(
+        PackedCTokenData<V>,
+        CompressedAccountMetaNoLamportsNoAddress,
+    )>,
+    proof: ValidityProof,
+    cpi_accounts: &CpiAccounts<'b, 'info>,
+    post_system_accounts: &[AccountInfo<'info>],
+    has_pdas: bool,
+    program_id: &Pubkey,
+) -> Result<(), ProgramError>
+where
+    V: CTokenSeedProvider<Accounts<'info> = A>,
+    A: 'info,
+{
+    let mut token_decompress_indices: Box<Vec<crate::instructions::DecompressFullIndices>> =
+        Box::new(Vec::with_capacity(ctoken_accounts.len()));
+    let mut token_signers_seed_groups: Vec<Vec<Vec<u8>>> =
+        Vec::with_capacity(ctoken_accounts.len());
+    let packed_accounts = post_system_accounts;
+
+    let authority = cpi_accounts
+        .authority()
+        .map_err(|_| ProgramError::MissingRequiredSignature)?;
+    let cpi_context = cpi_accounts
+        .cpi_context()
+        .map_err(|_| ProgramError::MissingRequiredSignature)?;
+
+    for (token_data, meta) in ctoken_accounts.into_iter() {
+        let owner_index: u8 = token_data.token_data.owner;
+        let mint_index: u8 = token_data.token_data.mint;
+        let mint_info = &packed_accounts[mint_index as usize];
+        let owner_info = &packed_accounts[owner_index as usize];
+
+        // Use trait method to get seeds (program-specific)
+        let (ctoken_signer_seeds, derived_token_account_address) = token_data
+            .variant
+            .get_seeds(accounts_for_seeds, remaining_accounts)?;
+
+        if derived_token_account_address != *owner_info.key {
+            msg!(
+                "derived_token_account_address: {:?}",
+                derived_token_account_address
+            );
+            msg!("owner_info.key: {:?}", owner_info.key);
+            return Err(ProgramError::InvalidAccountData);
+        }
+
+        let seed_refs: Vec<&[u8]> = ctoken_signer_seeds.iter().map(|s| s.as_slice()).collect();
```

Comment:
Since arrays of references have a small memory footprint `ArrayVec` with capacity 16 should not exceed stack bounds and be more CU efficient.
```suggestion
        let seed_refs: ArrayVec<&[u8]> = ctoken_signer_seeds.iter().map(|s| s.as_slice()).collect();
```